File: ios/PlaytomicFoundation/observable/Disposable.swift
//
//  Disposable.swift
//  Anemone SDK
//
//  Created by Angel Luis Garcia on 19/03/2019.
//  Copyright © 2019 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public final class Disposable {
    public let dispose: () -> Void

    init(_ dispose: @escaping () -> Void) {
        self.dispose = dispose
    }

    public func add(to disposal: IDisposal) {
        disposal.add(disposable: self)
    }
}

public final class Disposal: IDisposal {
    private var disposables: [Disposable]

    public init() {
        self.disposables = []
    }

    public func add(disposable: Disposable) {
        synchronized(self) {
            disposables.append(disposable)
        }
    }

    public func dispose() {
        synchronized(self) {
            disposables.forEach { $0.dispose() }
            disposables.removeAll()
        }
    }

    public func getDisposableSize() -> Int {
        disposables.count
    }
}

public protocol IDisposal {
    func add(disposable: Disposable)
    func dispose()
    func getDisposableSize() -> Int
}

File: android/app/src/main/java/com/playtomic/test/foundation/observable/Disposable.kt
package com.playtomic.foundation.observable

class Disposable(val dispose: (() -> Unit)) {

    fun add(to: IDisposal) {
        to.add(disposable = this)
    }
}

class Disposal : IDisposal {
    private val disposables: MutableList<Disposable> = mutableListOf()

    override fun add(disposable: Disposable) {
        disposables.add(disposable)
    }

    override fun dispose() {
        disposables.forEach { it.dispose() }
        disposables.clear()
    }

    override fun getDisposableSize() = disposables.size
}

interface IDisposal {
    fun add(disposable: Disposable)
    fun dispose()
    fun getDisposableSize(): Int
}
